x/merkledb/codec.go

```diff
@@ -91,16 +91,42 @@ type codecImpl struct {
 	varIntPool sync.Pool
 }
 
-func (c *codecImpl) encodeDBNode(n *dbNode) []byte {
-	var (
-		numChildren = len(n.children)
-		// Estimate size of [n] to prevent memory allocations
-		estimatedLen = estimatedValueLen + minVarIntLen + estimatedNodeChildLen*numChildren
-		buf          = bytes.NewBuffer(make([]byte, 0, estimatedLen))
-	)
+func (c *codecImpl) encodedDBNodeSize(n *dbNode) int {
+	// total the number of children pointers + bool indicating if it has a value + the value + the child entries for n.children
+	total := c.uintSize(uint64(len(n.children))) + boolLen
+	if n.value.HasValue() {
+		total += c.uintSize(uint64(len(n.value.Value()))) + len(n.value.Value())
+	}
+	// for each non-nil entry, we add the additional size of the child entry
+	for index, entry := range n.children {
+		total += c.childSize(index, entry)
+	}
+	return total
+}
 
+func (c *codecImpl) childSize(index byte, childEntry *child) int {
+	return c.uintSize(uint64(index)) + ids.IDLen + c.keySize(childEntry.compressedKey) + boolLen
+}
+
+// based on the current implementation of codecImpl.encodeUint which uses binary.PutUvarint
+func (*codecImpl) uintSize(value uint64) int {
+	// catch the call to log(0) and return early
+	if value == 0 {
+		return 1
+	}
+	// binary.PutUvarint repeatedly divides by 128 until the value is under 128,
+	// so count the number of times that occurs
+	return 1 + int(math.Log(float64(value))/log128)
```

Comment:
should we avoid floating point math here?

Reply:
it isn't the end of the world if it is off, but I could switch this to a looped version

Reply:
Done